### PR TITLE
Fixes a crash when some labels/boxes are NaN

### PR DESCRIPTION
**Pull request by BloodAxe**
Test dataset to reproduce. See the attached report generated
[Report.pdf](https://github.com/Deci-AI/data-gradients/files/14420160/Report.pdf)

If batch/sample with NaN is detected it is excluded from processing entirely. But it is tracked and added to list of errors which is not also written to PDF

```
import numpy as np
import torch

from data_gradients import DetectionAnalysisManager

train_samples = [
    # 4 images with labels
    (torch.randn((3, 100, 100), dtype=torch.float32), torch.randint(0, 10, (1, 5), dtype=torch.float32)),
    (torch.randn((3, 100, 100), dtype=torch.float32), torch.randint(0, 10, (2, 5), dtype=torch.float32)),
    (torch.randn((3, 100, 100), dtype=torch.float32), torch.randint(0, 10, (3, 5),

 dtype=torch.float32)),
    (torch.randn((3, 100, 100), dtype=torch.float32), torch.randint(0, 10, (4, 5), dtype=torch.float32)),
]
train_samples[3][1][0:2, 2] = np.nan

valid_samples = [
    # 4 images with labels
    (torch.randn((3, 100, 100), dtype=torch.float32), torch.randint(0, 10, (1, 5), dtype=torch.float32)),
    (torch.randn((3, 100, 100), dtype=torch.float32), torch.randint(0, 10, (2, 5), dtype=torch.float32)),
    (torch.randn((3, 100, 100), dtype=torch.float32), torch.randint(0, 10, (3, 5), dtype=torch.float32)),
    (torch.randn((3, 100, 100), dtype=torch.float32), torch.randint(0, 10, (4, 5), dtype=torch.float32)),
]
valid_samples[2][1][0, 2:4] = np.nan
manager = DetectionAnalysisManager(
    train_data=train_samples,
    val_data=valid_samples,
    report_title="Detection Test",
    class_names=["class_1", "class_2", "class_3", "class_4", "class_5", "class_6", "class_7", "class_8", "class_9",
                 "class_10"],
    batches_early_stop=None,
)
manager.run()
```

### CODE REVIEW

File: src/data_gradients/managers/abstract_manager.py [hunk 1]
```diff
@@ -95,38 +97,88 @@ def execute(self):
             "computer vision datasets. click here: https://hubs.ly/Q01XpHBT0"
         )
 
-        datasets_tqdm = tqdm(
-            zip_longest(self.train_samples_iterator, self.val_samples_iterator, fillvalue=None),
-            desc="Analyzing... ",
-            total=self.n_batches,
-        )
+        # datasets_tqdm = iter(tqdm(
```
Comment (ofrimasad) [inline src/data_gradients/managers/abstract_manager.py]: please remove commented out code

File: src/data_gradients/managers/abstract_manager.py [hunk 2]
```diff
@@ -95,38 +97,88 @@ def execute(self):
...
 
-        for i, (train_sample, val_sample) in enumerate(datasets_tqdm):
+        non_finite_train_samples = 0
+        non_finite_train_samples_example = ""
+        non_finite_val_samples = 0
+        non_finite_val_samples_example = ""
+
+        with tqdm(total=self.n_batches, desc="Analyzing... ") as datasets_tqdm:
+            i = 0
+            while True:
```
Comment (ofrimasad) [inline src/data_gradients/managers/abstract_manager.py:113]: why not `while i < self.batches_early_stop`

File: src/data_gradients/managers/abstract_manager.py [hunk 3]
```diff
@@ -95,38 +97,88 @@ def execute(self):
...
+                if i == self.batches_early_stop:
+                    self._stopped_early = True
+                    break
 
-            if i == self.batches_early_stop:
-                self._stopped_early = True
-                break
+                try:
+                    train_sample = None
```
Comment (ofrimasad) [inline src/data_gradients/managers/abstract_manager.py:115]: IS that required? wouldn't `next` return None if the iterator ended. 
If it is just an init, can you please move this line outside of the loop?

File: src/data_gradients/managers/abstract_manager.py [hunk 4]
```diff
@@ -95,38 +97,88 @@ def execute(self):
...
-                self._train_batch_size = self._train_iters_done
+                except StopIteration:
+                    train_sample = None
+                except NonFiniteValuesError as e:
+                    logger.error(f"Train sample {self._train_iters_done} contains non-finite values: {e}")
+                    non_finite_train_samples += 1
+                    non_finite_train_samples_example = str(e)
+
+                try:
+                    val_sample = None
```
Comment (ofrimasad) [inline src/data_gradients/managers/abstract_manager.py:135]: same question here

**Review by ofrimasad — COMMENTED** [review]
Nice solution.
A few minor comments

File: src/data_gradients/managers/abstract_manager.py (hunk 3 above)
Comment (BloodAxe) [reply src/data_gradients/managers/abstract_manager.py:115]: > wouldn't next return None if the iterator ended.
Nope, when iterator is exhausted it raises StopIteration exception and you should wait for it to know when it's ended.
So there are a few stopping criterias:
1) `batches_early_stop`
2) Both train and validation iterators are exhausted 

So yes, it is needed

File: src/data_gradients/managers/abstract_manager.py (hunk 2 above)
Comment (BloodAxe) [reply src/data_gradients/managers/abstract_manager.py:113]: The existing logic of the code uses `_stopped_early` flag that we set here to indicate in a report that only part of dataset was processed.
The logic is a bit croocked but I was not feeling confident enough to start refactoring additional piece of logic.

File: src/data_gradients/managers/abstract_manager.py (hunk 1 above)
Comment (BloodAxe) [reply src/data_gradients/managers/abstract_manager.py]: Fixed in e8f3e923454842e4a4f6c5446984f6c4100b2e67

File: src/data_gradients/managers/abstract_manager.py (hunk 3 above)
Comment (BloodAxe) [reply src/data_gradients/managers/abstract_manager.py:115]: In e8f3e923454842e4a4f6c5446984f6c4100b2e67 I've moved exit criterion checks to a single location at the end of the loop (easier to follow when all breaks are in a single place)

**Review by ofrimasad — APPROVED** [review]
LGTM
